runtime/common/JsonConvert.h

```diff
@@ -178,10 +197,236 @@ inline void from_json(const json &j, ExecutionContext &context) {
+    return ret_ptr;
+  }
+  // This shouldn't happen, but gracefully handle it if it does.
+  return std::make_unique<cudaq::optimizers::cobyla>();
+}
+
+// ----- cudaq::gradient serialization/deserialization support below
+
+enum class GradientEnum { CENTRAL_DIFF, FORWARD_DIFF, PARAMETER_SHIFT };
+NLOHMANN_JSON_SERIALIZE_ENUM(GradientEnum,
+                             {JSON_ENUM(GradientEnum, CENTRAL_DIFF),
+                              JSON_ENUM(GradientEnum, FORWARD_DIFF),
+                              JSON_ENUM(GradientEnum, PARAMETER_SHIFT)});
+
+inline GradientEnum get_gradient_type(const cudaq::gradient &p) {
+  if (dynamic_cast<const cudaq::gradients::central_difference *>(&p))
+    return GradientEnum::CENTRAL_DIFF;
+  if (dynamic_cast<const cudaq::gradients::forward_difference *>(&p))
+    return GradientEnum::FORWARD_DIFF;
+  if (dynamic_cast<const cudaq::gradients::parameter_shift *>(&p))
+    return GradientEnum::PARAMETER_SHIFT;
+  // This shouldn't happen, but handle it gracefully if it does.
+  return GradientEnum::CENTRAL_DIFF;
+}
+
+// These do not attempt to serialize or deserialize the quantum kernel
+// (ansatz_functor). That is intentional because those will be handled via the
+// RestRequest.code.
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(cudaq::gradients::central_difference, step);
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(cudaq::gradients::forward_difference, step);
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(cudaq::gradients::parameter_shift,
+                                   shiftScalar);
+
+inline void to_json(json &j, const cudaq::gradient &p) {
+  if (auto *central_difference =
+          dynamic_cast<const cudaq::gradients::central_difference *>(&p))
+    j = json(*central_difference);
+  else if (auto *forward_difference =
+               dynamic_cast<const cudaq::gradients::forward_difference *>(&p))
+    j = json(*forward_difference);
+  else if (auto *parameter_shift =
+               dynamic_cast<const cudaq::gradients::parameter_shift *>(&p))
+    j = json(*parameter_shift);
+}
+
+inline std::unique_ptr<cudaq::gradient>
+make_gradient_from_json(const nlohmann::json &j,
+                        const GradientEnum gradient_type) {
+  if (gradient_type == GradientEnum::CENTRAL_DIFF) {
+    auto ret_ptr = std::make_unique<cudaq::gradients::central_difference>();
+    from_json(j,
+              dynamic_cast<cudaq::gradients::central_difference &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (gradient_type == GradientEnum::FORWARD_DIFF) {
+    auto ret_ptr = std::make_unique<cudaq::gradients::forward_difference>();
+    from_json(j,
+              dynamic_cast<cudaq::gradients::forward_difference &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (gradient_type == GradientEnum::PARAMETER_SHIFT) {
+    auto ret_ptr = std::make_unique<cudaq::gradients::parameter_shift>();
+    from_json(j, dynamic_cast<cudaq::gradients::parameter_shift &>(*ret_ptr));
+    return ret_ptr;
+  }
```

Comment:
same question as https://github.com/NVIDIA/cuda-quantum/pull/1800/files#r1635394341

Reply:
Updated in e8a16e4db. I also removed the unnecessary dynamic_cast just to tidy it up.